### PR TITLE
[Feature][external catalog/lakesoul] LakeSoul Catalog support Filter Pushdown & Cdc data handling & S3 data access 

### LINKED ISSUE #37978: [Enhancement] LakeSoul Catalog support Filter Pushdown & Cdc data handling & S3 data access
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Support filter pushdown, cdc data handling and s3 data access for [LakeSoul](https://lakesoul-io.github.io/) Catalog

### Solution

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Ceng23333**
## Proposed changes

Issue Number: close #37978

### CODE REVIEW

File: fe/be-java-extensions/lakesoul-scanner/src/main/java/org/apache/doris/lakesoul/LakeSoulJniScanner.java [hunk 1]
```diff
@@ -72,19 +78,37 @@ public void open() throws IOException {
...
 
-        Schema schema = Schema.fromJSON(params.get(LakeSoulUtils.SCHEMA_JSON));
+        String options = params.getOrDefault(LakeSoulUtils.OPTIONS, "{}");
+        Map<String, String> optionsMap = new ObjectMapper().readValue(
+                options, new TypeReference<Map<String, String>>() {}
+        );
+        String base64Predicate = optionsMap.get(LakeSoulUtils.SUBSTRAIT_PREDICATE);
+        if (base64Predicate != null) {
+            Plan predicate = SubstraitUtil.decodeBase64String(base64Predicate);
+            LOG.info("push predicate={}", predicate);
```
Comment (morningman) [inline fe/be-java-extensions/lakesoul-scanner/src/main/java/org/apache/doris/lakesoul/LakeSoulJniScanner.java:88]: ditto

File: fe/fe-core/src/main/java/org/apache/doris/datasource/lakesoul/LakeSoulExternalTable.java [hunk 2]
```diff
@@ -150,6 +165,7 @@ public Optional<SchemaCacheValue> initSchema() {
         String tableSchema = tableInfo.getTableSchema();
         DBUtil.TablePartitionKeys partitionKeys = DBUtil.parseTableInfoPartitions(tableInfo.getPartitions());
         Schema schema;
+        LOG.info("tableSchema={}", tableSchema);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/lakesoul/LakeSoulExternalTable.java:168]: ditto

File: fe/fe-core/src/main/java/org/apache/doris/datasource/lakesoul/source/LakeSoulScanNode.java [hunk 3]
```diff
@@ -81,26 +132,32 @@ protected TFileFormatType getFileFormatType() throws UserException {
...
 
     @Override
     protected Map<String, String> getLocationProperties() throws UserException {
         return lakeSoulExternalTable.getHadoopProperties();
     }
 
+    @SneakyThrows
     @Override
     protected void setScanParams(TFileRangeDesc rangeDesc, Split split) {
+        LOG.info("{}", rangeDesc);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/lakesoul/source/LakeSoulScanNode.java:151]: ditto

File: fe/fe-core/src/main/java/org/apache/doris/datasource/lakesoul/source/LakeSoulScanNode.java [hunk 4]
```diff
@@ -126,24 +215,45 @@ public void setLakeSoulParams(TFileRangeDesc rangeDesc, LakeSoulSplit lakeSoulSp
     }
 
     public List<Split> getSplits() throws UserException {
+        LOG.info("getSplits with columnFilters={}", columnFilters);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/lakesoul/source/LakeSoulScanNode.java:218]: ditto

File: fe/fe-core/src/main/java/org/apache/doris/datasource/lakesoul/source/LakeSoulScanNode.java [hunk 5]
```diff
@@ -126,24 +215,45 @@ public void setLakeSoulParams(TFileRangeDesc rangeDesc, LakeSoulSplit lakeSoulSp
...
+                    LakeSoulUtils.applyPartitionFilters(
+                        allPartitionInfo,
+                        tableName,
+                        partitionArrowSchema,
+                        columnNameToRange
+                    );
+        } catch (IOException e) {
+            throw new UserException(e);
+        }
+        LOG.info("filteredPartitionInfo={}", filteredPartitionInfo);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/lakesoul/source/LakeSoulScanNode.java:236]: ditto

File: fe/fe-core/src/main/java/org/apache/doris/datasource/lakesoul/LakeSoulUtils.java [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+                            conjunctionFilter = SubstraitUtil.or(conjunctionFilter, expr);
+                        }
+                    }
+                }
+                return conjunctionFilter;
+            }
+        }
+    }
+
+    public static Expression rangeToSubstraitFilter(Field columnField, Range range) throws IOException {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/lakesoul/LakeSoulUtils.java:125]: I suggest to write unit test for these expr conversion logic.
You can refer to: IcebergPredicateTest.java

**Comment by morningman** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Comment by morningman** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Review by CalvinKirs — APPROVED** [review]

**Pull request merged**